internal/config/config.go

```diff
@@ -0,0 +1,168 @@
+// See sigs.k8s.io/secrets-store-csi-driver/provider/v1alpha1/service.pb.go
+type Config struct {
+	Parameters
+	TargetPath     string
+	FilePermission os.FileMode
+}
+
+// Parameters stores the parameters specified in the SecretProviderClass.
+// Top-level values that aren't strings are not directly deserialisable because
+// they are defined as literal string types:
+// https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/0ba9810d41cc2dc336c68251d45ebac19f2e7f28/apis/v1alpha1/secretproviderclass_types.go#L59
+//
+// So we just deserialise by hand to avoid complexity and two passes.
+type Parameters struct {
+	VaultRoleName                string
+	VaultAddress                 string
+	VaultKubernetesMountPath     string
+	KubernetesServiceAccountPath string
+	TLSConfig                    TLSConfig
+	Secrets                      []Secret
+}
+
+type TLSConfig struct {
+	VaultCAPEM         string
+	VaultCACertPath    string
+	VaultCADirectory   string
+	VaultTLSServerName string
+	VaultSkipTLSVerify bool
+}
+
+func (c TLSConfig) CertificatesConfigured() bool {
+	return c.VaultCAPEM != "" ||
+		c.VaultCACertPath != "" ||
+		c.VaultCADirectory != ""
+}
+
+type Secret struct {
+	ObjectName    string `yaml:"objectName"`
+	ObjectPath    string `yaml:"objectPath"`
+	ObjectVersion string `yaml:"objectVersion"`
+}
+
+func Parse(parametersStr, targetPath, permissionStr string) (Config, error) {
+	config := Config{
+		TargetPath: targetPath,
+	}
+
+	var err error
+	config.Parameters, err = parseParameters(parametersStr)
+	if err != nil {
+		return Config{}, err
+	}
+
+	err = json.Unmarshal([]byte(permissionStr), &config.FilePermission)
+	if err != nil {
+		return Config{}, err
+	}
+
+	err = config.Validate()
+	if err != nil {
+		return Config{}, err
+	}
+
+	return config, nil
+}
+
+func parseParameters(parametersStr string) (Parameters, error) {
+	var params map[string]string
+	err := json.Unmarshal([]byte(parametersStr), &params)
+	if err != nil {
+		return Parameters{}, err
+	}
+
+	var parameters Parameters
+	parameters.VaultRoleName = params["roleName"]
+	parameters.VaultAddress = params["vaultAddress"]
+	parameters.TLSConfig.VaultCAPEM = params["vaultCAPem"]
+	parameters.TLSConfig.VaultCACertPath = params["vaultCACertPath"]
+	parameters.TLSConfig.VaultCADirectory = params["vaultCADirectory"]
+	parameters.TLSConfig.VaultTLSServerName = params["vaultTLSServerName"]
+	parameters.VaultKubernetesMountPath = params["vaultKubernetesMountPath"]
+	parameters.KubernetesServiceAccountPath = params["KubernetesServiceAccountPath"]
+	if skipTLS, ok := params["vaultSkipTLSVerify"]; ok {
+		value, err := strconv.ParseBool(skipTLS)
+		if err == nil {
+			parameters.TLSConfig.VaultSkipTLSVerify = value
+		} else {
+			return Parameters{}, err
+		}
+	}
+
+	secretsYaml := params["objects"]
+	// TODO: There is an unnecessary map under objects, instead of just directly storing an array there.
+	// Deserialisation can be simplified a fair bit if we remove it.
```

Comment:
I plan to do this in a PR in the near future, but held off for now given the purpose and size of this PR.